core/src/common.cpp

```diff
@@ -0,0 +1,501 @@
+#include <irods/rcConnect.h>
+#include <irods/rcMisc.h> // For addKeyVal().
+#include <irods/rodsErrorTable.h>
+#include <irods/rodsKeyWdDef.h> // For KW_CLOSE_OPEN_REPLICAS.
+#include <irods/switch_user.h>
+#include <irods/ticketAdmin.h>
+
+#include <boost/any.hpp>
+#include <boost/algorithm/string.hpp>
+
+#include <curl/curl.h>
+#include <nlohmann/json.hpp>
+#include <spdlog/spdlog.h>
+#include <string>
+
+namespace irods::http
+{
+	auto fail(response_type& _response, status_type _status, const std::string_view _error_msg) -> response_type
+	{
+		_response.result(_status);
+		_response.set(field_type::server, s3::version::server_name);
+		_response.set(field_type::content_type, "application/json");
+		_response.body() = _error_msg;
+		_response.prepare_payload();
+		return _response;
+	} // fail
+
+	auto fail(response_type& _response, status_type _status) -> response_type
+	{
+		return fail(_response, _status, "");
+	} // fail
+
+	auto fail(status_type _status, const std::string_view _error_msg) -> response_type
+	{
+		response_type r{_status, 11};
+		return fail(r, _status, _error_msg);
+	} // fail
+
+	auto fail(status_type _status) -> response_type
+	{
+		response_type r{_status, 11};
+		return fail(r, _status, "");
+	} // fail
+
+	auto decode(const std::string_view _v) -> std::string
+	{
+		std::string result;
+		int decoded_length = -1;
+
+		if (auto* decoded = curl_easy_unescape(nullptr, _v.data(), static_cast<int>(_v.size()), &decoded_length);
+		    decoded) {
+			std::unique_ptr<char, void (*)(void*)> s{decoded, curl_free};
+			result.assign(decoded, decoded_length);
+		}
+		else {
+			result.assign(_v);
+		}
+
+		return result;
+	} // decode
+
+	auto encode(std::string_view _to_encode) -> std::string
+	{
+		char* tmp_encoded_data{curl_easy_escape(nullptr, _to_encode.data(), _to_encode.size())};
+		if (tmp_encoded_data == nullptr) {
+			return {std::cbegin(_to_encode), std::cend(_to_encode)};
+		}
+
+		std::string encoded_data{tmp_encoded_data};
+
+		curl_free(tmp_encoded_data);
+		return encoded_data;
+	} // encode
+
+	// TODO Create a better name.
+	auto to_argument_list(const std::string_view _urlencoded_string) -> std::unordered_map<std::string, std::string>
+	{
+		if (_urlencoded_string.empty()) {
+			return {};
+		}
+
+		std::unordered_map<std::string, std::string> kvps;
+
+		std::vector<std::string> tokens;
+		boost::split(tokens, _urlencoded_string, boost::is_any_of("&"));
+
+		std::vector<std::string> kvp;
+
+		for (auto&& t : tokens) {
+			boost::split(kvp, t, boost::is_any_of("="));
+
+			if (kvp.size() == 2) {
+				auto value = decode(kvp[1]);
+				boost::replace_all(value, "+", " ");
+				kvps.insert_or_assign(std::move(kvp[0]), value);
+			}
+			else if (kvp.size() == 1) {
+				kvps.insert_or_assign(std::move(kvp[0]), "");
+			}
+
+			kvp.clear();
+		}
+
+		return kvps;
+	} // to_argument_list
+
+	auto get_url_path(const std::string& _url) -> std::optional<std::string>
+	{
+		namespace log = irods::http::log;
```

Comment:
Do not use the name `log` here as it is already defined in libstdc++.

This applies to all instances of `namespace log` in this PR.

Reply:
Will replace with `logging`, just like in your HTTP API PR.

https://github.com/irods/irods_client_http_api/pull/254

Reply:
Done.